subscription: Add real time sync for user-just-deactivated case.

YJDave opened the pull request:
Currently, stream subscriptions aren't getting updated without
hard reload when user is deactivated in realm.

Fix this issue by updating stream subscription widgets on user
deactivation event.

Fixes #5623

**Testing Plan:** <!-- How have you tested? -->
Manual testing:
* Open any stream subscription(say X) in incognito window
* Login as realm admin in another window
* Deactivate any user subscribed to stream X
* Check if user is removed from stream subscribers in another window or not
* You can also check in admin window, by opening stream subscribers tab

timabbott commented on static/js/stream_data.js:
```diff
@@ -345,6 +345,17 @@ exports.add_subscriber = function (stream_name, user_id) {
 
+exports.remove_deactivated_user_from_all_streams = function (user) {
+    (stream_info.values()).forEach(function (stream) {
+        if (exports.user_is_subscribed(stream.name, user.email)) {
```
Might be better to do a small `people.js` refactor to make a `exports.user_is_subscribed_by_id` rather than going through the user's email; we are trying to migrate away from unnecessary use of email addresses in the Zulip JS codebase.

timabbott commented on static/js/stream_data.js:
```diff
@@ -345,6 +345,17 @@ exports.add_subscriber = function (stream_name, user_id) {
+        if (exports.user_is_subscribed(stream.name, user.email)) {
+            exports.remove_subscriber(stream.name, user.user_id);
+            var sub = exports.get_sub(stream.name);
+            stream_edit.rerender_subscribers_list(sub);
```
This is really expensive; I think we only need to rerender the subscriber list for a currently being viewed stream (not all of them).

YJDave commented on frontend_tests/node_tests/compose_fade.js:
```diff
@@ -26,11 +26,11 @@ var bob = {
 };
 
-people.add(me);
+people.add_in_realm(me);
```
Reason of change is: `add` function was only adding user in all people dict, where as `add_in_realm` function adds user in active people dict, which is used for condition in function `would_receive_message`.

YJDave commented on static/js/subs.js:
```diff
@@ -225,14 +225,11 @@ exports.update_settings_for_subscribed = function (sub) {
 
-    var active_stream = exports.active_stream();
-    if (active_stream !== undefined && active_stream.id === sub.stream_id) {
+    if (sub.render_subscribers && stream_edit.is_sub_settings_active(sub)) {
```
Subscriptions are already render here, so we don't need call for `add_me_to_member_list`. And as this functions was not used in any other codebase, I removed it.

YJDave commented:
@timabbott Thanks for reviewing! It's updated, though it's circle tests are running. Addressed all comments, tested manually with two browser opened and with close/open subscriptions modal.

timabbott commented on static/js/stream_edit.js:
```diff
@@ -108,11 +97,6 @@ exports.update_stream_description = function (sub) {
-exports.prepend_subscriber = function (sub_row, email) {
-    var list = get_subscriber_list(sub_row);
-    list.prepend(format_member_list_elem(email));
-};
```
I think the purpose of this was to make sure that you appear at the top if you just subscribed yourself...

YJDave replied:
Yeah, but it wasn't working. Added it's fix.

timabbott commented on static/js/stream_edit.js:
```diff
@@ -58,6 +58,9 @@ exports.hide_sub_settings = function (sub) {
 exports.show_sub_settings = function (sub) {
+    if (!exports.is_sub_settings_active(sub)) {
+        return;
+    }
```
@YJDave can you explain why this fix is correct in more detail?  I feel like based on the name, `exports.show_sub_settings` here feels like something that should be used to display the target tab, though the code doesn't seem to match that use case ... what is this function actually for?

YJDave replied:
This function is used to show subscription related settings in stream setting page, i.e. notifications, email address etc. So when current user subscribed to stream we show/slides in these all settings. This can return error if there is currently no stream setting page open. In that case we don't need to render anything so just return.

timabbott commented:
I posted one comment, otherwise this looks great!

YJDave commented on static/js/subs.js:
```diff
@@ -259,7 +259,7 @@ exports.update_settings_for_subscribed = function (sub) {
     }
 
-    if (stream_edit.is_sub_settings_active(sub)) {
+    if (stream_edit.is_sub_settings_active(sub) && sub.invite_only) {
```
We don't want rerender subscribers, cause that will rerender subscribers in alphabetical order. But in case of private stream we want to, cause when user subscribe to private stream, there is no existing subscriber list so we need to rerender whole subscriber list. So if user is subscribed to private stream(when some other private stream's subscribed user add them) we may not have currently subscribed user on top of list, if user has stream subscription modal opened.

timabbott replied:
This doesn't seem like the right fix.  I'm merging it for now since it's better than the status quo, but if we think the correct behavior is to prepend the user, we should make that work, private stream or no (we could do that easily by extending the state).  The other option would be to always sort the current user at the top, which might be a better solution...

Can you open an issue for this question and mention Rishi to get an opinion on whether we should just always sort the current user to the top?

YJDave replied:
Sure, opened #9027

YJDave commented:
@timabbott Thanks for reviewing! Updated!

timabbott commented:
Merged, thanks @YJDave!  If you have time, it'd be super valuable to do a round of 2-browser QA on the stream settings system and try to find new tracebacks or other issues.  The main new thing I've found playing with this recently is that if you get added to a private stream, the "preview" button doesn't appear in the UI.

See also the comment on follow-up, above.

As a sidenote, I think we would potentially benefit from investing in adding comments for why we're doing a few of the funny things we're doing (e.g. why we do we need `is_sub_already_present`?).  I think that could help a lot for both our future efforts maintaining this as well as those of anyone who gets involved in this subsystem in the future.

YJDave commented:
Thanks for reviewing :) Doing QA, soon submit PR for comments part.
